client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +155,116 @@ public void start(Listener listener) {
+                        !this.scheduledFuture.isDone() &&
+                        this.scheduledFuture.getDelay(TimeUnit.SECONDS) > FAILURE_RETRY_TIMEOUT_SECONDS) {
+                    // Cancel the existing schedule and advance the discovery process.
+                    this.scheduledFuture.cancel(true);
+                    this.scheduledFuture = this.scheduledExecutor.schedule(
+                            this::getControllers, FAILURE_RETRY_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+                }
+            }
+        }
+
+        /**
+         * The controller discovery API invoker.
+         * This refreshes the list of controller addresses to be used by the gRPC transport.
+         * The discovery process will be rescheduled in the end after a delay which is calculated based on whether
+         * the controller addresses have been fetched successfully or not.
+         */
+        @Synchronized
         private void getControllers() {
+            log.info("Attempting to refresh the controller server endpoints");
             try {
-                final ServerResponse controllerServerList =
-                        this.client.getControllerServerList(ServerRequest.getDefaultInstance());
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                final ResolvedServerInfoGroup serverInfoGroup;
+                if (this.enableDiscovery) {
+                    // Make an RPC call to the bootstrapped controller servers to fetch all active controllers.
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server hostnames to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
```

Comment:
I guess synchronization was the wrong word for it. I meant the cost associated with using multiple threads, i.e. for creating substreams and merging the results. Need to understand which threads get used for this operation and costs associated with it (using the fork/join pool? and any waits associated with it). Without parallelism we are sure the operations will use the current thread and doesn't need additional resources. Need to explore this further to avoid any surprises